src/analysis/utils/parsing_with_cscope.py

```
@@ -0,0 +1,270 @@
+        boundary = candidates[lo]
+    else:
+        boundary = n   # no break found → take everything
+
+    # 5) Slice out everything from start to boundary (including blanks/braces)
+    return lines[start_ln - 1 : boundary]
+
+
+def get_global_definitions_by_lines(path, lines=None):
+    """
+    If `lines` is provided, it must be the same list that was written to `path`.
+    Otherwise we read & split the file once.
+    """
+    if lines is None:
+        with open(path, "r") as f:
+            src = f.read().splitlines(keepends=True)
+    else:
+        src = lines
+    out = [""] * len(src)
+
+    idx = Index.create()
+    # You can adjust -I flags here if you have more include dirs
+    tu  = idx.parse(path, args=["-std=c11", "-nostdinc"])
+
+    def visit(node):
+        loc = node.location
+        if not loc.file or loc.file.name != path:
+            return
+        if node.kind in {
+            CursorKind.VAR_DECL,
+            CursorKind.STRUCT_DECL,
+            CursorKind.ENUM_DECL,
+            CursorKind.MACRO_DEFINITION
+        } and node.semantic_parent.kind == CursorKind.TRANSLATION_UNIT:
+            s, e = node.extent.start.line, node.extent.end.line
+            for j in range(s-1, e):
+                out[j] = src[j].rstrip("\n")
+
+    for child in tu.cursor.get_children():
+        visit(child)
+
+    return out
+
+
+
+def parse_and_modify_with_cscope(content,
+                                 modified_line_numbers,
+                                 temp_dir,
+                                 file_name):
+    """
+    - content: full source text
+    - modified_line_numbers: list[int]
+    - temp_dir: writes ONLY preprocessed_<file_name> + cscope DB
+    - file_name: e.g. "foo.c"
+    """
+    os.makedirs(temp_dir, exist_ok=True)
+
+    # 1) Preprocess _in memory_
+    preproc_path, pp_lines, preproc_lines = preprocess_in_memory(content, temp_dir, file_name)
+    assert len(preproc_lines) == len(content.splitlines(keepends=True)), \
```

It's better to raise an error yourself than use `assert`